R/scales-.R

```diff
@@ -61,13 +61,44 @@ ScalesList <- ggproto("ScalesList", NULL,
     scale[[1]]
   },
 
+  train = function(self, data, drop = FALSE) {
+    if (length(self$scales) == 0) {
+      return()
+    }
+    lapply(data, self$train_df, drop = drop)
+  },
+
   train_df = function(self, df, drop = FALSE) {
     if (empty(df) || length(self$scales) == 0) {
       return()
     }
     lapply(self$scales, function(scale) scale$train_df(df = df))
   },
 
+  map = function(self, data, layers) {
+    if (length(self$scales) == 0) {
+      return(data)
+    }
+    aesthetics <- lapply(self$scales, `[[`, "aesthetics")
+
+    known <- unique(unlist(c(
+      lapply(data, colnames),
+      lapply(layers, function(x) names(x$computed_mapping))
```

Comment:
Sometimes you can have a mapping like `colour = after_scale(alpha(fill, 0.3))` that isn't instantiated as a column in the data yet. In order to not warn about these, we need to access the layer's mapping.